PSI_Tracker.js

```diff
@@ -0,0 +1,681 @@
+      batch[i].label,
+      batch[i].url,
+      batch[i].device,
+      batch[i].origin,
+    ];
+    let debug = "";
+    // Indicate that we worked on that row in the Config tab
+    sheet.getRange(batch[i].id + 1, COLUMN_STATUS + 1).setValue("✅");
+    // Read and parse the data for that specific cakk
+    const data = fetch[i];
+    let content = {};
+    // Try to parse the answer from the API
+    try {
+      content = JSON.parse(data);
+    } catch (e) {
+      // If an error occurs (ex: HTML returned instead of JSON), throw it here
+      sheet.getRange(batch[i].id + 1, COLUMN_STATUS + 1).setValue("🟥");
+      sheet
+        .getRange(batch[i].id + 1, COLUMN_STATUS + 1)
+        .setNote(
+          "Error " + e.toString() + "\n\nFull error:\n" + JSON.stringify(data)
+        );
+      debug +=
+        "🟥 " +
+        "Error " +
+        e.toString() +
+        "\n\nFull error:\n" +
+        JSON.stringify(data) +
+        " | ";
+      fields_for_debug.push(debug);
+      DEBUG_SHEET.appendRow(fields_for_debug);
+    }
+    Logger.log(content);
+    // If an error occured, flagged it in the Config tab, under the Status
+    if (content["error"]) {
+      sheet.getRange(batch[i].id + 1, COLUMN_STATUS + 1).setValue("❌");
+      sheet
+        .getRange(batch[i].id + 1, COLUMN_STATUS + 1)
+        .setNote("Error " + JSON.stringify(content["error"]));
+      debug += "❌ " + "Error " + JSON.stringify(content["error"]) + " | ";
+      fields_for_debug.push(debug);
+      DEBUG_SHEET.appendRow(fields_for_debug);
+    } else {
+      // Build a dic that will contain all the extracted information
+      let output = {
+        Date: new Date().toISOString().slice(0, 10),
+        Label: batch[i].label,
+        URL: batch[i].url,
+        Device: batch[i].device,
+        "URL / Origin": batch[i].origin,
+      };
+      // Iterate through all the fields and eval the function to get their respective values, adding them to the dic
+      for (let j = 0; j < fields.length; j++) {
+        try {
+          // Origin field is used in some calls to know if there was no URL-level data and we reverted to Origin level
+          let origin = batch[i].origin;
+          let green_domains_ = green_domains;
+          // Eval the formula, for example "content["lighthouseResult"]["lighthouseVersion"]"
+          let value = eval(fields[j].data);
+          // Add it to our dic
+          output[fields[j].field] = value;
+        } catch (e) {
+          // If error, write it in placce -> we ultimately decided to leave it blank instead as it was breaking dashboarding down the line
+          // output[fields[j].field] = "Error " + JSON.stringify(e.toString())
+          output[fields[j].field] = "";
+          debug +=
+            "❗️ Error on " +
+            fields[j].field +
+            JSON.stringify(e.toString()) +
+            " | ";
+        }
+      }
+
+      // STEP 3. We build the new row based on what we read on the Results first line
```

Comment:
I still don't like the amount of repetition here. Would it be possible to use the if to set `sheet_results` and `results_header` and then have a default case for everything but CrUX history? Having everything separate seems to defeat the purpose of `runBatch`. These could have been separate functions instead of creating the aliases above.

Reply:
yes, will try to refractor this